### ISSUE TITLE
When scenario samplers are executing concurrently DB persistence fails

### ISSUE THREAD

**Issue by yasassri** [wso2/testgrid#424]
**Description:**

If Test scenario samplers are executing parallel the DB updating fails with the following error.

```
ERROR {} Detected problem in Listener. java.lang.RuntimeException: Error occurred when persisting test case.
	at org.wso2.testgrid.automation.executor.JMeterResultCollector.sampleOccurred(JMeterResultCollector.java:69)
	at org.apache.jmeter.threads.ListenerNotifier.notifyListeners(ListenerNotifier.java:65)
	at org.apache.jmeter.threads.JMeterThread.notifyListeners(JMeterThread.java:891)
	at org.apache.jmeter.threads.JMeterThread.executeSamplePackage(JMeterThread.java:528)
	at org.apache.jmeter.threads.JMeterThread.processSampler(JMeterThread.java:424)
	at org.apache.jmeter.threads.JMeterThread.run(JMeterThread.java:255)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.wso2.testgrid.dao.TestGridDAOException: Error occurred when persisting entity in database.
	at org.wso2.testgrid.dao.repository.AbstractRepository.persist(AbstractRepository.java:73)
	at org.wso2.testgrid.dao.repository.TestCaseRepository.persist(TestCaseRepository.java:53)
	at org.wso2.testgrid.dao.uow.TestCaseUOW.persistTestCase(TestCaseUOW.java:69)
	at org.wso2.testgrid.automation.executor.JMeterResultCollector.sampleOccurred(JMeterResultCollector.java:67)
	... 6 more
Caused by: java.lang.IllegalStateException: 
Exception Description: Transaction is currently active
	at org.eclipse.persistence.internal.jpa.transaction.EntityTransactionImpl.begin(EntityTransactionImpl.java:45)
	at org.wso2.testgrid.dao.repository.AbstractRepository.persist(AbstractRepository.java:65)
	... 9 more
```

**Suggested Labels:**
Bug

**Comment by kasunbg** [comment]
@yasassri  Can you explain what is exactly meant by running scenario samplers in parallel? Are you executing multiple _run-testplans_ at once? Please mention the steps to re-produce this.

Thanks.

**Comment by yasassri** [comment]
This can be reproduced by adding multiple thread groups to the Scenario JMeter script and executing.

**Comment by azinneera** [comment]
This is happening because the EntityManager is not thread-safe by default. This has been solved in Spring. We should be able to solve this by @PersistenceContext EntityManager in Spring. WDYT?

**Comment by kasunbg** [comment]
+1 to bring in Spring to manage this. Rather than the @PersistenceContext, we can inject the @PersistenceUnit and get the EntityManagerFactory. 

Please raise a separate ticket for this. We need a immediate fix for the original issue. :)

**Comment by azinneera** [comment]
Created #466 for this.
